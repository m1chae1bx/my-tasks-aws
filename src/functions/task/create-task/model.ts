import { JSONSchemaType, ajv } from "@libs/generic/util";

export interface CreateTaskRequest {
  pathParameters: {
    listId: string;
  };
  body: {
    name: string;
    desc?: string;
    dueDate?: string;
  };
}

export const createTaskRequestSchema: JSONSchemaType<CreateTaskRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        listId: {
          type: "string",
        },
      },
      required: ["listId"],
    },
    body: {
      type: "object",
      properties: {
        name: {
          type: "string",
          maxLength: 200,
        },
        desc: {
          type: "string",
          nullable: true,
          maxLength: 400,
        },
        dueDate: {
          type: "string",
          nullable: true,
          format: "date-time",
        },
      },
      required: ["name"],
    },
  },
  required: ["pathParameters", "body"],
};

export const validateRequest = ajv.compile(createTaskRequestSchema);
