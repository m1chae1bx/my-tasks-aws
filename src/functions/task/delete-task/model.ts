import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface DeleteTaskRequest {
  pathParameters: {
    listId: string;
    taskId: string;
  };
}

export const deleteTaskRequestSchema: JSONSchemaType<DeleteTaskRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        listId: {
          type: "string",
        },
        taskId: {
          type: "string",
        },
      },
      required: ["listId", "taskId"],
    },
  },
  required: ["pathParameters"],
};

export const validateRequest = ajv.compile(deleteTaskRequestSchema);
