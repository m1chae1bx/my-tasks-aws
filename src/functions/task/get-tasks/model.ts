import { GetTasksQuery } from "/opt/nodejs/task.model";
import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface GetTasksRequest {
  pathParameters: {
    listId: string;
  };
  queryStringParameters?: GetTasksQuery;
}

export const getTasksRequestSchema: JSONSchemaType<GetTasksRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        listId: {
          type: "string",
        },
      },
      required: ["listId"],
    },
    queryStringParameters: {
      type: "object",
      properties: {
        name: {
          type: "string",
          nullable: true,
        },
        dueDate: {
          type: "string",
          enum: ["today", "tomorrow", "upcoming", "overdue", "unplanned"],
          nullable: true,
        },
        today: {
          type: "string",
          nullable: true,
          format: "date-time",
        },
        includeCompleted: {
          type: "string",
          enum: ["true"],
          nullable: true,
        },
      },
      nullable: true,
      if: { properties: { dueDate: { type: "string" } } },
      then: { required: ["today"] },
    },
  },
  required: ["pathParameters"],
};

export const validateRequest = ajv.compile(getTasksRequestSchema);
