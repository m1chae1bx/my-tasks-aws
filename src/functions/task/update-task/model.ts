import { JSONSchemaType, ajv } from "@libs/generic/util";

export interface UpdateTaskRequest {
  pathParameters: {
    listId: string;
    taskId: string;
  };
  body: {
    id: string;
    name: string;
    isCompleted?: boolean;
    desc?: string;
    dueDate?: string;
  };
}

export const updateTaskRequestSchema: JSONSchemaType<UpdateTaskRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        listId: {
          type: "string",
        },
        taskId: {
          type: "string",
        },
      },
      required: ["listId", "taskId"],
    },
    body: {
      type: "object",
      properties: {
        id: {
          type: "string",
        },
        name: {
          type: "string",
          maxLength: 200,
        },
        isCompleted: {
          type: "boolean",
          nullable: true,
        },
        desc: {
          type: "string",
          nullable: true,
          maxLength: 400,
        },
        dueDate: {
          type: "string",
          nullable: true,
          format: "date-time",
        },
      },
      required: ["id", "name"],
    },
  },
  required: ["pathParameters", "body"],
  if: {
    properties: {
      pathParameters: {
        type: "object",
        properties: {
          taskId: {
            type: "string",
          },
        },
      },
    },
  },
  then: {
    properties: {
      body: {
        type: "object",
        properties: {
          id: {
            const: { $data: "/pathParameters/taskId" },
          },
        },
      },
    },
  },
};

export const validateRequest = ajv.compile(updateTaskRequestSchema);
