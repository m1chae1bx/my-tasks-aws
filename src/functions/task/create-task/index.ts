import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { validateRequest } from "./model";
import { Task } from "/opt/nodejs/task.model";
import { genericErrorHandler } from "/opt/nodejs/util";

export const handler = async (
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> => {
  try {
    const request = {
      pathParameters: event.pathParameters,
      body: event.body ? JSON.parse(event.body) : {},
    };

    if (!validateRequest(request)) {
      return {
        statusCode: 400,
        body: JSON.stringify({
          message: "Bad request",
          errors: validateRequest.errors,
        }),
      };
    }

    const { name, dueDate, desc } = request.body;
    const { listId } = request.pathParameters;
    const task = new Task(
      listId,
      name,
      false,
      dueDate ? new Date(dueDate) : undefined,
      desc
    );

    const id = await task.save();
    const response = { message: `Task '${name}' was created successfully`, id };
    return {
      statusCode: 200,
      body: JSON.stringify(response),
    };
  } catch (error) {
    return genericErrorHandler(
      error,
      "An error occurred while creating the task. Please try again later."
    );
  }
};
