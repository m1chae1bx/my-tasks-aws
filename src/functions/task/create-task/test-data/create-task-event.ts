import { baseAPIGatewayProxyEvent } from "@test-utils/base-objects";

export const createTaskEvent = {
  ...baseAPIGatewayProxyEvent,
  pathParameters: {
    listId: "test-list-id",
  },
  body: JSON.stringify({
    name: "test-task-name",
    desc: "test-task-desc",
    dueDate: "2020-01-01T00:00:00.000Z",
  }),
};
