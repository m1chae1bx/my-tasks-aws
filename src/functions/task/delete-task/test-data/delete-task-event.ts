import { baseAPIGatewayProxyEvent } from "@test-utils/base-objects";

export const deleteTaskEvent = {
  ...baseAPIGatewayProxyEvent,
  pathParameters: {
    listId: "test-list-id",
    taskId: "test-task-id",
  },
};
