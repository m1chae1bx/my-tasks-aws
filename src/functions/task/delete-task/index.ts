import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { validateRequest } from "./model";
import { Task } from "@libs/task";
import { genericErrorHandler } from "@libs/generic/util";

export const handler = async (
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> => {
  try {
    const request = {
      pathParameters: event.pathParameters,
    };
    if (!validateRequest(request)) {
      return {
        statusCode: 400,
        body: JSON.stringify({
          message: "Bad request",
          errors: validateRequest.errors,
        }),
      };
    }

    const { listId, taskId } = request.pathParameters;
    await Task.delete(taskId, listId);

    const response = { message: `Task ${taskId} was deleted successfully` };
    return {
      statusCode: 200,
      body: JSON.stringify(response),
    };
  } catch (error) {
    return genericErrorHandler(
      error,
      `An error occurred while deleting the task. Please try again later.`
    );
  }
};
