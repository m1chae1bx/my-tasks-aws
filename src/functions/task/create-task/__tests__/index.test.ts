import { ajvError } from "@test-utils/base-objects";
import { APIGatewayProxyResult } from "aws-lambda";
import { handler } from "..";
import { validateRequest } from "../model";
import { createTaskEvent } from "../test-data/create-task-event";
import { Task } from "/opt/nodejs/task.model";
import { genericErrorHandler } from "/opt/nodejs/util";

jest.mock("../model", () => ({
  validateRequest: jest.fn(),
}));

jest.mock("/opt/nodejs/task.model", () => ({
  Task: jest.fn(),
}));

jest.mock("/opt/nodejs/util", () => ({
  genericErrorHandler: jest.fn(),
}));

describe("create-task", () => {
  describe("happy path", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(true);
      (Task as unknown as jest.Mock).mockReturnValueOnce({
        save: jest.fn().mockResolvedValueOnce("test-task-id"),
      });
      result = await handler(createTaskEvent);
    });

    it("should return a 200 response", () => {
      expect(result.statusCode).toBe(200);
    });

    it("should return a success message", () => {
      expect(result.body).toMatchSnapshot();
      expect(result.body).toContain("test-task-id");
    });
  });

  describe("sad path - invalid request", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(false);
      validateRequest.errors = ajvError;
      result = await handler(createTaskEvent);
    });

    it("should return a 400 response", () => {
      expect(result.statusCode).toBe(400);
    });

    it("should return the validation errors", () => {
      expect(result.body).toMatchSnapshot();
    });
  });

  describe("sad path - error while saving task", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(true);
      (Task as unknown as jest.Mock).mockReturnValueOnce({
        save: jest.fn().mockRejectedValueOnce(new Error("test-error")),
      });
      (genericErrorHandler as jest.Mock).mockReturnValueOnce({
        statusCode: 500,
        body: "test error",
      });
      result = await handler(createTaskEvent);
    });

    it("should return a 500 response", () => {
      expect(result.statusCode).toBe(500);
    });

    it("should return the error message", () => {
      expect(result.body).toMatchSnapshot();
    });
  });
});
