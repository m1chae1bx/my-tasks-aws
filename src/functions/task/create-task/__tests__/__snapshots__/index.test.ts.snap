// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create-task happy path should return a success message 1`] = `"{\\"message\\":\\"Task 'test-task-name' was created successfully\\",\\"id\\":\\"test-task-id\\"}"`;

exports[`create-task sad path - error while saving task should return the error message 1`] = `"test error"`;

exports[`create-task sad path - invalid request should return the validation errors 1`] = `"{\\"message\\":\\"Bad request\\",\\"errors\\":[{\\"keyword\\":\\"\\",\\"instancePath\\":\\"\\",\\"schemaPath\\":\\"\\",\\"params\\":{},\\"message\\":\\"test error\\"}]}"`;
